docs/ADRs/draft/adr-4_new-train-keys.md

```diff
@@ -0,0 +1,86 @@
+---
+adr: 4
+title: Securing validator and participants keys
+author: Silas Davis (@silasdavis), Sean Young (@seanyoung), Greg Hill (@gregdhill)
+discussions-to: https://chat.hyperledger.org/channel/burrow-contributors
+status: Draft
+type: Non-Standards Track
+category: Security, Keys management
+created: 2019-08-21
+---
+
+## Securing validator and participants keys
+
+This ADR introduces how we will ensure that private key matter is protected.
+
+## Motivation
+
+A burrow node can be a server for many clients; however, at the moment many clients use mempool signing
+which requires the burrow node to hold the private keys for all participants, and all client needs to know
+is the address of the account and access to the node's grpc port. This is not secure.
```

Comment:
I think we need to unpack 'this is not secure'. It's not like we didn't think about the implications of mempool signing. Officially mempool signing is meant for co-located signing where Burrow is secured ambiently. This is not inherently insecure.

However I think the problem is this massively encourages being insecure by default since:

- Burrow needs to be exposed to internet to function - if you use mempool signing you need to be careful to expose peer port but not GRPC port. We don't really document this anywhere.

- Burrow feels like a one-to-many server from a client's perspective so mempool signing tends to introduce an assumption of trust between participants sharing a node which is probably an anti-pattern. The path of least resistance means all of your clients can use each others keys. Our context is dominated by a single security domain - so what if our API servers share a key? However I think this encourages bad practice - it is not ideal even in our case.

- They way the implementation currently works includes the validator key which was unlikely to be the explicit intention.

We need to consider with this proposal whether we address some of these by pushing them into keys. Thinking about it makes me think we probably need to default to having a separate keys service. I wonder if we should drop support for running keys in Burrow...